server/pulp/plugins/migration/standard_storage_path.py

```diff
@@ -0,0 +1,374 @@
+        :type unit: Unit
+        """
+        item = Item(unit.plan, unit.id, unit.storage_path, unit.new_path)
+        self.items[unit.storage_path] = item
+
+    def reset(self):
+        """
+        Reset the batch by clearing added units.
+        """
+        self.items = {}
+
+    def _relink(self):
+        """
+        Update (relink) symlinks to reflect the storage path of migrated units.
+        Foreach symlink found, find the *new* path using the plan.  Then,
+        re-create the symlink with the updated target.
+        """
+        root = Migration.publish_dir()
+        for path, directories, files in os.walk(root):
+            for name in chain(files, directories):
+                abs_path = os.path.join(path, name)
+                if not os.path.islink(abs_path):
+                    continue
+                target = os.readlink(abs_path)
+                unit = self.items.get(target)
+                if not unit:
+                    continue
+                os.unlink(abs_path)
+                os.symlink(unit.new_path, abs_path)
+
+    def _migrate(self):
+        """
+        Migrate the units referenced in the batch.
+          1. Move the content files.
+          2. Update the unit in the DB.
+        """
+        for item in self.items.values():
+            Unit.migrate(item.plan, item.unit_id, item.storage_path, item.new_path)
+
+    def __call__(self):
+        """
+        Execute the batch.
+          1. Update published links to reference the new path.
+          2. Move the content to the new path.
+          2. Update _storage_path on the unit and saved.
+        """
+        if self.items:
+            self._relink()
+            self._migrate()
+            self.reset()
+
+    def __len__(self):
+        return len(self.items)
+
+
+class Plan(object):
+    """
+    The Unit migration plan contains all that is necessary
+    to migrate a specific type of content unit.  This includes the collection,
+    the list of unit keys (needed to calculate the new storage path) and an
+    indication of whether the storage path references a file or directory.
+
+    :ivar collection: A DB collection.
+    :type collection: pymongo.collection.Collection
+    :ivar key_fields: List of fields that make up the unit key.
+    :type key_fields: tuple
+    :ivar join_leaf: Join the existing path leaf (filename) to the new path.
+    :type join_leaf: bool
+    """
+
+    def __init__(self, collection, key_fields, join_leaf=True):
+        """
+        :param collection: A DB collection.
+        :type collection: pymongo.collection.Collection
+        :param key_fields: List of fields that make up the unit key.
+        :type key_fields: tuple
+        :param join_leaf: Join the existing path leaf (filename) to the new path.
+        :type join_leaf: bool
+        """
+        self.collection = collection
+        self.key_fields = key_fields
+        self.join_leaf = join_leaf
+
+    @property
+    def base_fields(self):
```

Comment:
This is less obvious because it's decorated as a property.  The reason for making this a property was so that sub-classes could include additional fields to be pulled.  I ran into this when doing the migration for the `python` plugin.  Now that you have pointed this out, I'm not crazy about using a property for this.  It would be safer to provide a way for the sub-class specify additional fields instead of redefining the entire list (with the override).  I'll revise this.